src/DIRAC/FrameworkSystem/Service/TokenManagerHandler.py

```diff
@@ -211,39 +211,40 @@ def export_getToken(
             return result
         idpObj = result["Value"]
 
-        if userGroup and (result := idpObj.getGroupScopes(userGroup))["OK"]:
+        if userGroup and (result := idpObj.getGroupScopes(userGroup)):
             # What scope correspond to the requested group?
-            scope = list(set((scope or []) + result["Value"]))
+            scope = list(set((scope or []) + result))
 
         # Set the scope
         idpObj.scope = " ".join(scope)
 
-        # Let's check if there are corresponding tokens in the cache
-        cacheKey = (username, idpObj.scope, audience, identityProvider)
+        # Let's check if there is a corresponding token in the cache
+        cacheKey = (userName, idpObj.scope, audience, identityProvider)
         if self.__tokensCache.exists(cacheKey, requiredTimeLeft):
             # Well we have a fresh record containing a Token object
             token = self.__tokensCache.get(cacheKey)
             # Let's check if the access token is fresh
             if not token.is_expired(requiredTimeLeft):
                 return S_OK(token)
-            # It seems that it is no longer valid for us, but whether there is a refresh token?
-            if token.get("refresh_token"):
-                # Okay, so we can try to refresh tokens
-                if (result := idpObj.refreshToken(token["refresh_token"]))["OK"]:
-                    # caching new tokens
-                    self.__tokensCache.add(
-                        cacheKey,
-                        result["Value"].get_claim("exp", "refresh_token") or self.DEFAULT_RT_EXPIRATION_TIME,
-                        result["Value"],
-                    )
-                    return result
-                self.log.verbose(f"Failed to get token with cached tokens: {result['Message']}")
-                # Let's try to revoke broken token
-                idpObj.revokeToken(token["refresh_token"])
 
         err = []
-        # The cache did not help, so let's make an exchange token
-        result = Registry.getDNForUsername(username)
+        # Do we have an oidc-agent configured ? If yes, try to use it
+        if userName in idpObj.parameters.get("OIDCAgentUsers", []):
+            result = idpObj.getTokenFromAgent(userName, requiredTimeLeft=requiredTimeLeft)
+
+            if result["OK"]:
+                # caching new tokens
+                self.__tokensCache.add(
+                    cacheKey,
+                    result["Value"].get_claim("exp", "refresh_token") or self.DEFAULT_RT_EXPIRATION_TIME,
+                    result["Value"],
+                )
+                return result
+            else:
+                err.append(result["Message"])
+
+        # No luck so far, let's refresh the token stored in the database
```

Comment:
I'm struggling to understand where this is using the REST interface of IAM/CheckIn to get the tokens. Am I missing somthing?

Reply:
IIUC, here: https://github.com/DIRACGrid/DIRAC/pull/6750/files/d94157958af74dffd0dca4abea0ef82cf6106342#diff-d582aa662e6992b4d896c3050cfb033627550a2c5ea1f0b67f6f54298c0647beR262

If `idpObj` is an instance of `OAuth2IdProvider`, then it calls `authlib.OAuthClient.refreshToken()` which makes the REST request.
(`OAuth2IdProvider` inherits from `OAuth2Session`, which inherits from `authlib.OAuth2Session`, which inherits from `authlib.OAuth2Client`)